crates/common/src/config.rs

```diff
@@ -1,8 +1,192 @@
+                                alias.into(),
+                                Utf8PathBuf::from(path),
+                                parameters,
+                            ));
+                        } else {
+                            diagnostics.push(ConfigDiagnostic::InvalidDependencyDescription);
+                        }
+                    }
+                    _ => diagnostics.push(ConfigDiagnostic::InvalidDependencyDescription),
+                }
+            }
+        }
+
+        Ok(Self {
+            metadata,
+            dependencies,
+            diagnostics,
+        })
+    }
+
+    pub fn based_dependencies(&self, base_url: &Url) -> Vec<BasedDependency> {
+        self.dependencies
+            .iter()
+            .map(|dependency| dependency.based(base_url))
+            .collect()
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
 pub struct IngotMetadata {
     pub name: Option<SmolStr>,
     pub version: Option<Version>,
 }
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum DependencyDescription {
+    Path(Utf8PathBuf),
+    PathWithParameters {
+        path: Utf8PathBuf,
+        parameters: DependencyParameters,
+    },
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Dependency {
+    pub alias: SmolStr,
+    pub description: DependencyDescription,
+}
+
+impl Dependency {
+    pub fn path(alias: SmolStr, path: Utf8PathBuf) -> Self {
+        Self {
+            alias,
+            description: DependencyDescription::Path(path),
+        }
+    }
+
+    pub fn path_with_arguments(
+        alias: SmolStr,
+        path: Utf8PathBuf,
+        parameters: DependencyParameters,
+    ) -> Self {
+        Self {
+            alias,
+            description: DependencyDescription::PathWithParameters { path, parameters },
+        }
+    }
+
+    pub fn based(&self, base_url: &Url) -> BasedDependency {
+        match &self.description {
+            DependencyDescription::Path(path) => BasedDependency {
+                alias: self.alias.clone(),
+                parameters: DependencyParameters::default(),
+                url: base_url.join_directory(path).unwrap(),
+            },
+            DependencyDescription::PathWithParameters {
+                path,
+                parameters: _,
+            } => BasedDependency {
+                alias: self.alias.clone(),
+                parameters: DependencyParameters::default(),
+                url: base_url.join_directory(path).unwrap(),
+            },
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct BasedDependency {
+    pub alias: SmolStr,
+    pub parameters: DependencyParameters,
+    pub url: Url,
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
+pub struct DependencyParameters {
+    pub name: Option<SmolStr>,
+    pub version: Option<Version>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum ConfigDiagnostic {
```

Comment:
there should be a tracked method on ingots for getting these and they should be reported on the CLI